File: 39.combinationSum/combinationSum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        dfs(candidates, target, res, path, 0);
        return res;
    }
    
    private void dfs(int[] candidates, int target, List<List<Integer>> res, List<Integer> path, int index) {
        //System.out.println(index + ", " + path);
        if (target == 0) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = index; i < candidates.length; i++) {
            if (target >= candidates[i]) {
                path.add(candidates[i]);
                dfs (candidates, target - candidates[i], res, path, i);
                path.remove(path.size() - 1);
            }
        }
    }
}

File: 39.combinationSum/combinationSum.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int> > res;
        vector<int> path;
        dfs(candidates, target, path, res, 0);
        return res;
    }
    void dfs(vector<int>& candidates, int target, vector<int>& path, vector<vector<int> >& res, int start) {
        if (target == 0) {
            res.push_back(path);
        }
        for (int i = start; i < candidates.size(); i++) {
            if (candidates[i] <= target) {
                path.push_back(candidates[i]);
                dfs(candidates, target - candidates[i], path, res, i);
                path.pop_back();
            }
        }
    }
};

int main()
{
    Solution s;
    int a[4] = {2,3,6,7};
    vector<int> vec(a, a+4);
    vector<vector<int> > res = s.combinationSum(vec, 7);
    for (vector<int> vec : res) {
        for (int i : vec)
            cout << i << " ";
        cout << endl;
    }
    return 0;
}
